src/data_source/sql.rs

```diff
@@ -0,0 +1,987 @@
+    type BlockStream = BoxStream<'static, BlockQueryData<Types>>;
+
+    type LeafRange<'a, R> = BoxStream<'a,  QueryResult<LeafQueryData<Types, I>>>
+    where
+        Self: 'a,
+        R: RangeBounds<usize> + Send;
+    type BlockRange<'a, R>= BoxStream<'a,  QueryResult<BlockQueryData<Types>>>
+    where
+        Self: 'a,
+        R: RangeBounds<usize> + Send;
+
+    async fn get_leaf<ID>(&self, _id: ID) -> QueryResult<LeafQueryData<Types, I>>
+    where
+        ID: Into<LeafId<Types, I>> + Send + Sync,
+    {
+        todo!()
+    }
+    async fn get_block<ID>(&self, _id: ID) -> QueryResult<BlockQueryData<Types>>
+    where
+        ID: Into<BlockId<Types>> + Send + Sync,
+    {
+        todo!()
+    }
+
+    async fn get_leaf_range<R>(&self, _range: R) -> QueryResult<Self::LeafRange<'_, R>>
+    where
+        R: RangeBounds<usize> + Send,
+    {
+        todo!()
+    }
+    async fn get_block_range<R>(&self, _range: R) -> QueryResult<Self::BlockRange<'_, R>>
+    where
+        R: RangeBounds<usize> + Send,
+    {
+        todo!()
+    }
+
+    async fn get_block_with_transaction(
+        &self,
+        _hash: TransactionHash<Types>,
+    ) -> QueryResult<(BlockQueryData<Types>, TransactionIndex<Types>)> {
+        todo!()
+    }
+
+    async fn get_proposals(
+        &self,
+        _proposer: &EncodedPublicKey,
+        _limit: Option<usize>,
+    ) -> QueryResult<Vec<LeafQueryData<Types, I>>> {
+        todo!()
+    }
+    async fn count_proposals(&self, _proposer: &EncodedPublicKey) -> QueryResult<usize> {
+        todo!()
+    }
+
+    async fn subscribe_leaves(&self, _height: usize) -> QueryResult<Self::LeafStream> {
+        todo!()
+    }
+    async fn subscribe_blocks(&self, _height: usize) -> QueryResult<Self::BlockStream> {
+        todo!()
+    }
+}
+
+#[async_trait]
+impl<Types, I> UpdateAvailabilityData<Types, I> for SqlDataSource
+where
+    Types: NodeType,
+    I: NodeImplementation<Types>,
+    Block<Types>: QueryableBlock,
+{
+    type Error = postgres::error::Error;
+
+    async fn insert_leaf(&mut self, _leaf: LeafQueryData<Types, I>) -> Result<(), Self::Error>
+    where
+        Deltas<Types, I>: Resolvable<Block<Types>>,
+    {
+        todo!()
+    }
+
+    async fn insert_block(&mut self, _block: BlockQueryData<Types>) -> Result<(), Self::Error> {
+        todo!()
+    }
+}
+
+#[async_trait]
+impl StatusDataSource for SqlDataSource {
+    type Error = postgres::error::Error;
+
+    async fn block_height(&self) -> Result<usize, Self::Error> {
+        todo!()
+    }
+    async fn mempool_info(&self) -> Result<MempoolQueryData, Self::Error> {
+        todo!()
+    }
+    async fn success_rate(&self) -> Result<f64, Self::Error> {
+        todo!()
+    }
+    async fn export_metrics(&self) -> Result<String, Self::Error> {
+        todo!()
+    }
+}
+
+impl UpdateStatusData for SqlDataSource {
+    fn metrics(&self) -> Box<dyn Metrics> {
+        todo!()
+    }
+}
+
+pub struct Transaction<'a> {
```

Comment:
lol. An earlier iteration of this required `Transaction` to be `'static`, so I implemented this wrapper around client. Then apparently I got rid of that restriction and added the lifetime back in, but now the whole reason for having a custom `Transaction` type is gone...I think I can switch it back to `tokio_postgres::Transaction` now

Reply:
Oh no, I remember why this was necessary. https://github.com/EspressoSystems/hotshot-query-service/pull/265#discussion_r1406455649

`tokio_postgres::Transaction` executes a `BEGIN` command every time it is created and a `COMMIT` every time it is dropped. But to avoid a self-reference I create and drop these `Transaction` wrappers all the time, and only create/commit a transaction in the database when `commit_version`/`revert_version` is called. So I think this wrapper is still necessary sadly

Reply:
Added a comment explaining this in [a1423b1](https://github.com/EspressoSystems/hotshot-query-service/pull/265/commits/a1423b142a29f313cbe9a4310725d9f5972333e8)